Add helper service for pictures/images

m-vo opened the pull request:
| Q                | A
| -----------------| ---
| Fixed issues     | none
| Docs PR or issue | todo

This is a draft for a service that eases working with pictures/images. The idea is to get an easy to use API that wraps around the `PictureFactory` / `ImageFactory` and eliminates the tedious tasks of retrieving and setting the file, meta data and paths (root dir, static url).

#### Scenarios
```php
// get values
$sources = $imageHelper->createPicture($uuid, '_my_size')->getSources();
$path = $imageHelper->createImage(123, 3)->getPath();

// render a picture template
$pictureTemplate->setData(
  $imageHelper->createPicture('/path/to/my/file.jpg', [200, 300])->getData()
);
```

We could do without the DTOs but imho this would result in a worse DX.


#### Naming
I'm really unsure about the naming though. In the image domain lots of generating services and classes are already there (`PictureFactory`, 'PictureGenerator', `Picture`, ...) that might be confused with this service. I called it `Helper` for now. (Intuitively I would have assumed this functionality in the factory classes, but I'm not sure if it's a good idea to add it there and add more [optional?] constructor parameters. Maybe?)


#### Background
Right now the only alternative is (mis)using `Controller::addImageToTemplate()` or the `image`/`picture` insert tag for this job. In fact these two parts of the code base could be rewritten to use a common service as well.

This would be helpful for #1528. With this we could even easily add an analogous template function that does the same (e.g. `$this->createPicture(...)`).

m-vo committed:
add image helper service

aschempp reviewed (commented):
I like the idea of a helper to generate pictures or images. I don't think we want a `Helper` namespace or additional `Helper` classes though. When looking at `Controller::addImageToTemplate`, the helper method does a lot more than just "wrap the data in an object". It also parses additional properties from `tl_content`.

`ImageHelper` seems right to me, there's also a `SecurityHelper` in Symfony that wraps security services for example.

m-vo committed:
remove Helper namespace + return arrays instead of DTOs

m-vo commented:
> I like the idea of a helper to generate pictures or images. I don't think we want a Helper namespace or additional Helper classes though. 

Agreed. I switched to returning arrays (with constants as keys) instead of DTOs and removed the extra namespace. @aschempp Is this what you were thinking about?

Toflar commented:
I would also like to have something like that in the core 👍 Would certainly require @ausi's input :)

fritzmg reviewed (changes requested):
Nice work :). I do would want to change one thing though: don't make the file path dependent on the file being present in the DBAFS. In my opinion it can be useful to process files from anywhere and also from within `files/` regardless of whether the file is actually present in the DBAFS or not. `getFile` could instead return a `\Contao\File` object and only when fetching the meta attributes (e.g. right now for `getAltAttribute`) it should look whether the file is actually present in the DBAFS and fetch the data accordingly.

fritzmg commented on core-bundle/src/Resources/config/services.yml:
```diff
@@ -334,6 +334,15 @@ services:
         public: true
 
+    contao.image.helper:
+        class: Contao\CoreBundle\Image\Helper\ImageHelper
```
```suggestion
    Contao\CoreBundle\Image\ImageHelper:
```
🤓

m-vo commented:
> In my opinion it can be useful to process files from anywhere 

We could also allow `File` as an input btw.

But maybe we should instead just accept a path in this service and have another one that acts as an (image) resource locator and can deal with the various inputs.

@ausi How does the (deferred) resizing work with resources that aren't accessible through a public path. Are they still served through the asset management (even if not resized at all)?

m-vo commented:
Memo to myself: consider supporting inline vs deferred creation.

ausi commented:
> How does the (deferred) resizing work with resources that aren't accessible through a public path. Are they still served through the asset management (even if not resized at all)?

Depends on the settings. If you specify a target path for a single image resize, no deferred resizing (and asset management) is used. If you enable the `skipIfDimensionsMatch` option, the source path of the image is used when no resize is necessary. The image factory itself doesn’t know if the path to a source image is public or not.

aschempp commented on core-bundle/src/Image/ImageHelper.php:
```diff
@@ -0,0 +1,113 @@
+        $this->framework = $framework;
+
+        $request = $requestStack->getCurrentRequest();
+        $this->locale = null !== $request ? $request->getLocale() : '';
```
should this use the system default local (`en` usually configured in the framework) instead of an empty value?

aschempp reviewed (commented):
also, this should probably boot the framework if I want to use the service in Symfony?

m-vo committed:
initialize framework

m-vo committed:
CS

m-vo committed:
fallback to default locale

m-vo committed:
refactor to use filepaths instead of forcing DBAFS models

m-vo commented:
Ok, I made some small changes (i.e. allow using file paths independently of the DBAFS).

Could you guys check, if that's roughly going in the direction you're after and/or provide some input? If so I'd like to get this PR ready (tests, the same for images, ...) soon to finally get #1528 out :smile: .

/cc @aschempp @ausi @Toflar @fritzmg

ausi commented on core-bundle/src/Image/ImageHelper.php:
```diff
@@ -0,0 +1,147 @@
+        $this->framework = $framework;
+
+        $request = $requestStack->getCurrentRequest();
+        $this->locale = null !== $request ? $request->getLocale() : $defaultLocale;
```
Doesn’t this make the service stateful? What if the locale changes between the construction of the service and a call to `createPicture()`? Wouldn’t it be better to pass the locale directly to `createPicture`?

m-vo replied:
You're right I guess. We need a way to pass the locale anyway (e.g. when running on CLI). Maybe we should still fallback to the request locale / default locale so that one does not need to provide it in most cases? Wdyt?

bytehead replied:
You can just get the actual request from the request stack in the method you use the locale (just move it away from the constructor).

m-vo replied:
Just saw, that `addImageToTemplate()` refers to `$objPage->language` / `$objPage->rootFallbackLanguage`. To recreate this behavior we should imo use the request to check for a page and follow the same logic if `createPicture()` wasn't called with an explicit locale.

ausi commented:
I like the general concept so far. But I think we should add all the functionality of `addImageToTemplate()` so that it can be a complete replacement.

Toflar commented:
I agree with @ausi. I also think it shouldn't be named `ImageHelper`. It's a service, every service helps you do something and is thus a "helper" 😊 Moreover, it already resides in the `Image` namespace making it redundant. 
I have a hard time finding a good name for something that "provides shortcuts for image handling" too though. What about giving it a name from the real world where you'd actually work with images? Like `Studio`? `Lab`?
I don't know but `Helper` just seems random 😄

m-vo commented:
> I also think it shouldn't be named ImageHelper

I totally agree, but couldn't find anything better, yet, that wasn't already used and fitted the scope (i.e. not sounding like photo editing/image manipulation). Maybe we need to brainstorm a bit more. 

> Studio? Lab?

`Workshop`? `Creator`? `Maker`? `Builder`? `Tools`? `Frontend`? `Atelier`? :smile:

m-vo committed:
allow setting a locale

m-vo committed:
WIP

m-vo commented:
closing in favor #1753
